lib/vfs/src/union_fs.rs

```diff
@@ -0,0 +1,426 @@
+#![allow(dead_code)]
+#![allow(unused)]
+use crate::*;
+use std::borrow::Cow;
+use std::ops::Add;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::AtomicU32;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::sync::Weak;
+#[allow(unused_imports, dead_code)]
+use tracing::{debug, error, info, trace, warn};
+
+pub type TempHolding = Arc<Mutex<Option<Arc<Box<dyn FileSystem>>>>>;
+
+#[derive(Debug)]
+pub struct MountPoint {
+    pub path: String,
+    pub name: String,
+    pub fs: Option<Arc<Box<dyn FileSystem>>>,
+    pub weak_fs: Weak<Box<dyn FileSystem>>,
+    pub temp_holding: TempHolding,
+    pub should_sanitize: bool,
+    pub new_path: Option<String>,
+}
+
+impl Clone for MountPoint {
+    fn clone(&self) -> Self {
+        Self {
+            path: self.path.clone(),
+            name: self.name.clone(),
+            fs: None,
+            weak_fs: self.weak_fs.clone(),
+            temp_holding: self.temp_holding.clone(),
+            should_sanitize: self.should_sanitize,
+            new_path: self.new_path.clone(),
+        }
+    }
+}
+
+impl MountPoint {
+    pub fn fs(&self) -> Option<Arc<Box<dyn FileSystem>>> {
+        match &self.fs {
+            Some(a) => Some(a.clone()),
+            None => self.weak_fs.upgrade(),
+        }
+    }
+
+    fn solidify(&mut self) {
+        if self.fs.is_none() {
+            self.fs = self.weak_fs.upgrade();
+        }
+        {
+            let mut guard = self.temp_holding.lock().unwrap();
+            let fs = guard.take();
+            if self.fs.is_none() {
+                self.fs = fs;
+            }
+        }
+    }
+
+    fn strong(&self) -> Option<StrongMountPoint> {
+        self.fs().map(|fs| StrongMountPoint {
+            path: self.path.clone(),
+            name: self.name.clone(),
+            fs,
+            should_sanitize: self.should_sanitize,
+            new_path: self.new_path.clone(),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct StrongMountPoint {
+    pub path: String,
+    pub name: String,
+    pub fs: Arc<Box<dyn FileSystem>>,
+    pub should_sanitize: bool,
+    pub new_path: Option<String>,
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct UnionFileSystem {
+    pub mounts: Vec<MountPoint>,
```

Comment:
No, it allows you to mix multiple filesystems in one filesystem at different mount points.

Reply:
Sure, but mem_fs already supports nested file systems, so this could easily be implemented with mem_fs and a few changes.

Probably better left for a future cleanup though.

Reply:
Probably mem_fs doesn't support redirecting file paths (i.e. `/my_mount/my_file.txt` -> `/my_file.txt`), but that's just my guess.